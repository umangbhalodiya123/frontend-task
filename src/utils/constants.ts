import { Dashboard, Info, Person, Settings } from '@mui/icons-material'

export const API_URL = 'https://api.mockaroo.com/api/502602f0'

export const API_KEY = process.env.NEXT_PUBLIC_API_KEY

export const CREDENTIALS = [
  {
    id: '1',
    name: 'John Doe',
    email: 'admin@mail.com',
    password: 'Admin@123',
  },
]

export const NAVBAR_WIDTH = 240

export const NAVBAR_ITEMS = [
  {
    name: 'Dashboard',
    path: '/',
    icon: Dashboard,
  },
  {
    name: 'About',
    path: '/about',
    icon: Info,
  },
  {
    name: 'Profile',
    path: '/profile',
    icon: Person,
  },
  {
    name: 'Settings',
    path: '/settings',
    icon: Settings,
  },
]
