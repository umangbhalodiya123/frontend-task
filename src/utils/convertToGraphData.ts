import { Employee } from '@/types'

export const convertToGraphData = (data: Employee[], key: keyof Employee) => {
  const graphData: { label: string; value: number }[] = []

  const labels = [...new Set(data.map((employee) => employee[key]))].toSorted(
    (a, b) =>
      typeof a === 'number' && typeof b === 'number'
        ? a - b
        : typeof a === 'string' && typeof b === 'string'
        ? a.localeCompare(b)
        : 0
  )

  labels.forEach((label) => {
    const avgPerformanceRating =
      data
        .filter((employee) => employee[key] === label)
        .reduce((acc, curr) => acc + curr.performance_rating, 0) /
      data.filter((employee) => employee[key] === label).length

    graphData.push({
      label: String(label),
      value: avgPerformanceRating,
    })
  })

  return graphData
}
