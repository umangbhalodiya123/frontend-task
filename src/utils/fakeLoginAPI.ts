import { LoginRequestValues, User } from '../types'
import { CREDENTIALS } from '.'

export const login = (values?: LoginRequestValues, timeout = 700) => {
  return new Promise<{ success: boolean; user?: User }>((resolve, reject) => {
    setTimeout(() => {
      const user = CREDENTIALS.find(
        (credential) =>
          credential.email === values?.email &&
          credential.password === values?.password
      )
      if (user) {
        resolve({ success: true, user })
      } else {
        reject({ success: false })
      }
    }, timeout)
  })
}
