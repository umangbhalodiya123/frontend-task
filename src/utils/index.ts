export * from './theme'
export * from './constants'
export * from './storage'
export * from './convertToGraphData'
export * from './fakeLoginAPI'
export * from './mocks'
