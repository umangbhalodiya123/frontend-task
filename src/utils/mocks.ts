export const employees = [
  {
    id: 1,
    name: 'Ruby Goadby',
    gender: 'Male',
    age: 24,
    job_title: 'Teacher',
    department: 'Sales',
    performance_rating: 5,
    attendance_percentage: 87.87,
    overtime_hours: 7.55,
    team_size: 11,
    communication_skills: 'good',
    employee_status: 'on leave',
  },
  {
    id: 2,
    name: 'Terencio Feeney',
    gender: 'Male',
    age: 24,
    job_title: 'Social Worker',
    department: 'Legal',
    performance_rating: 4,
    attendance_percentage: 87.2,
    overtime_hours: 19.8,
    team_size: 3,
    communication_skills: 'average',
    employee_status: 'active',
  },
  {
    id: 3,
    name: 'Orazio Libbis',
    gender: 'Male',
    age: 23,
    job_title: 'Nurse Practicioner',
    department: 'Services',
    performance_rating: 3,
    attendance_percentage: 96.54,
    overtime_hours: 15.61,
    team_size: 9,
    communication_skills: 'good',
    employee_status: 'active',
  },
  {
    id: 4,
    name: 'Meara Jewell',
    gender: 'Female',
    age: 26,
    job_title: 'Analyst Programmer',
    department: 'Research and Development',
    performance_rating: 8,
    attendance_percentage: 36.57,
    overtime_hours: 5.66,
    team_size: 7,
    communication_skills: 'good',
    employee_status: 'active',
  },
  {
    id: 5,
    name: 'Farlee Solomonides',
    gender: 'Male',
    age: 23,
    job_title: 'Human Resources Assistant IV',
    department: 'Business Development',
    performance_rating: 7,
    attendance_percentage: 57.37,
    overtime_hours: 17.96,
    team_size: 9,
    communication_skills: 'poor',
    employee_status: 'active',
  },
  {
    id: 6,
    name: 'Brandice Kilbee',
    gender: 'Female',
    age: 21,
    job_title: 'VP Sales',
    department: 'Product Management',
    performance_rating: 2,
    attendance_percentage: 31.24,
    overtime_hours: 15.78,
    team_size: 6,
    communication_skills: 'good',
    employee_status: 'active',
  },
  {
    id: 7,
    name: 'Helsa Grealy',
    gender: 'Agender',
    age: 39,
    job_title: 'Recruiting Manager',
    department: 'Training',
    performance_rating: 9,
    attendance_percentage: 73.98,
    overtime_hours: 13.99,
    team_size: 15,
    communication_skills: 'excellent',
    employee_status: 'active',
  },
  {
    id: 8,
    name: 'Kimmie Bowart',
    gender: 'Female',
    age: 23,
    job_title: 'Assistant Professor',
    department: 'Services',
    performance_rating: 10,
    attendance_percentage: 41.31,
    overtime_hours: 19.8,
    team_size: 3,
    communication_skills: 'average',
    employee_status: 'active',
  },
  {
    id: 9,
    name: 'Desmond Bligh',
    gender: 'Male',
    age: 31,
    job_title: 'Software Test Engineer II',
    department: 'Services',
    performance_rating: 4,
    attendance_percentage: 4.27,
    overtime_hours: 4.57,
    team_size: 9,
    communication_skills: 'poor',
    employee_status: 'on leave',
  },
  {
    id: 10,
    name: 'Dana Schuelcke',
    gender: 'Male',
    age: 37,
    job_title: 'Safety Technician II',
    department: 'Training',
    performance_rating: 5,
    attendance_percentage: 42.8,
    overtime_hours: 19.09,
    team_size: 17,
    communication_skills: 'average',
    employee_status: 'on leave',
  },
]
