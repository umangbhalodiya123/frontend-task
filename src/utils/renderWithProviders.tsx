import React, { PropsWithChildren } from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { store } from '@/store'

export function renderWithProviders(ui: React.ReactElement) {
  const Wrapper = ({ children }: PropsWithChildren) => (
    <Provider store={store}>{children}</Provider>
  )

  return {
    store,
    ...render(ui, { wrapper: Wrapper }),
  }
}
