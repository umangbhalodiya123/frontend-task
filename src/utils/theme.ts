'use client'
import { Roboto } from 'next/font/google'
import { createTheme } from '@mui/material/styles'
import { deepOrange, orange } from '@mui/material/colors'

const roboto = Roboto({
  weight: ['300', '400', '500', '700'],
  subsets: ['latin'],
  display: 'swap',
})

export const theme = createTheme({
  palette: {
    mode: 'dark',
    primary: deepOrange,
    secondary: orange,
  },
  typography: {
    fontFamily: roboto.style.fontFamily,
  },
})
