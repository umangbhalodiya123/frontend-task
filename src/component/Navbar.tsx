'use client'
import { NAVBAR_ITEMS, NAVBAR_WIDTH } from '@/utils'
import { ChevronLeft } from '@mui/icons-material'
import {
  Divider,
  Drawer as MuiDrawer,
  IconButton,
  List,
  ListItemButton,
  ListItemIcon,
  ListItemText,
  Toolbar,
} from '@mui/material'
import Link from 'next/link'
import { styled } from '@mui/material/styles'

const Drawer = styled(MuiDrawer, {
  shouldForwardProp: (prop) => prop !== 'open',
})(({ theme, open }) => ({
  '& .MuiDrawer-paper': {
    position: 'relative',
    whiteSpace: 'nowrap',
    width: NAVBAR_WIDTH,
    transition: theme.transitions.create('width', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen,
    }),
    boxSizing: 'border-box',
    ...(!open && {
      overflowX: 'hidden',
      transition: theme.transitions.create('width', {
        easing: theme.transitions.easing.sharp,
        duration: theme.transitions.duration.leavingScreen,
      }),
      width: theme.spacing(7),
      [theme.breakpoints.up('sm')]: {
        width: theme.spacing(9),
      },
    }),
  },
}))

type P = Readonly<{
  open: boolean
  toggleDrawer: () => void
}>

export const Navbar = ({ open, toggleDrawer }: P) => {
  return (
    <Drawer variant="permanent" open={open}>
      <Toolbar
        sx={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'flex-end',
          px: [1],
        }}
      >
        <IconButton onClick={toggleDrawer}>
          <ChevronLeft />
        </IconButton>
      </Toolbar>

      <Divider />

      <List component="nav">
        {NAVBAR_ITEMS.map((item) => (
          <ListItemButton key={item.name} LinkComponent={Link} href={item.path}>
            <ListItemIcon>
              <item.icon />
            </ListItemIcon>
            <ListItemText primary={item.name} />
          </ListItemButton>
        ))}
      </List>
    </Drawer>
  )
}
