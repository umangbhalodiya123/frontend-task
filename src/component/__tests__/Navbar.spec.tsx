import { renderWithProviders } from '@/utils/renderWithProviders'
import { Navbar } from '../Navbar'

describe('Navbar', () => {
  it('renders and matches snapshot', () => {
    const { container } = renderWithProviders(
      <Navbar open={false} toggleDrawer={() => {}} />
    )
    expect(container).toMatchSnapshot()
  })
})
