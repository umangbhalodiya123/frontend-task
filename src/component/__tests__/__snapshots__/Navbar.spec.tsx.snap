// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders and matches snapshot 1`] = `
<div>
  <div
    class="MuiDrawer-root MuiDrawer-docked css-aq1gdx-MuiDrawer-docked"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft css-12i7wg6-MuiPaper-root-MuiDrawer-paper"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1dd0suk-MuiToolbar-root"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ChevronLeftIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <nav
        class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
      >
        <a
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          href="/"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="DashboardIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 13h8V3H3zm0 8h8v-6H3zm10 0h8V11h-8zm0-18v6h8V3z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              Dashboard
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
        <a
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          href="/about"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="InfoIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 15h-2v-6h2zm0-8h-2V7h2z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              About
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
        <a
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          href="/profile"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="PersonIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4m0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              Profile
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
        <a
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          href="/settings"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SettingsIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              Settings
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
      </nav>
    </div>
  </div>
</div>
`;
